### PR TITLE
Run analysis after profile

### PR DESCRIPTION
**Pull request by nv-braf**
Refactored code in analyzer
Fixed bugs in metrics manger and result table manager
Updated relevant L0 tests so they only run profile

### CODE REVIEW

File: model_analyzer/analyzer.py [hunk 1]
```diff
@@ -64,7 +70,8 @@ def __init__(self, config, server, state_manager, checkpoint_required):
         self._result_manager = ResultManager(config=config,
                                              state_manager=self._state_manager)
 
-    def profile(self, client, gpus):
+    def profile(self, client: Union[TritonGRPCClient, TritonHTTPClient],
+                gpus: list, mode: str, verbose: bool):
```
Comment (nv-hwoo) [inline model_analyzer/analyzer.py:74]: Type hint also supports `List` that allows us to specify the object types inside the list as well.

File: model_analyzer/entrypoint.py [hunk 2]
```diff
@@ -260,7 +260,10 @@ def main():
                                 server,
                                 state_manager,
                                 checkpoint_required=False)
-            analyzer.profile(client=client, gpus=gpus)
+            analyzer.profile(client=client,
+                             gpus=gpus,
+                             mode=args.mode,
+                             verbose=bool(args.verbose))
```
Comment (nv-hwoo) [inline model_analyzer/entrypoint.py:266]: I think `args.verbose` is already boolean type.

File: model_analyzer/analyzer.py [hunk 3]
```diff
@@ -223,6 +241,42 @@ def _profile_models(self):
...
                     self._state_manager.save_checkpoint()
 
+    def _analyze_models(self):
+        # TODO: Until we get rid of analysis we need to copy some values from profile
+        self._config._fields["analysis_models"] = self._config._fields[
+            "profile_models"]
+
+        gpu_info = self._state_manager.get_state_variable('MetricsManager.gpus')
+        if not gpu_info:
+            gpu_info = {}
```
Comment (nv-hwoo) [inline model_analyzer/analyzer.py:251]: Is `gpu_info` used here?

File: model_analyzer/analyzer.py [hunk 4]
```diff
@@ -98,11 +105,22 @@ def profile(self, client, gpus):
...
+        if (self._state_manager.exiting()):
+            sys.exit(1)
+
+        # TODO: TMA-792: This won't be needed once the Results class is used in profile
+        self._analyze_models()
+
+        self._create_summary_tables(verbose)
+
+        # TODO: TMA-790: Skip this step if the CLI option indicates the user doesn't want summary reports
+        self._create_summary_reports(mode)
```
Comment (nv-hwoo) [inline model_analyzer/analyzer.py:117]: Does the order matter here? Just making sure since `analyze` seems to run these in the opposite order.

File: model_analyzer/result/result_table_manager.py [hunk 5]
```diff
@@ -74,7 +75,8 @@ def tabulate_results(self):
 
         # Fill rows in descending order
         for model in self._result_manager.get_model_names():
-            result_heap = self._result_manager.get_model_sorted_results(model)
+            result_heap = deepcopy(
+                self._result_manager.get_model_sorted_results(model))
```
Comment (nv-hwoo) [inline model_analyzer/result/result_table_manager.py:79]: Do we need to copy result_heap?

File: model_analyzer/analyzer.py [hunk 6]
```diff
@@ -98,11 +105,22 @@ def profile(self, client, gpus):
 
         self._profile_models()
 
+        if (self._state_manager.exiting()):
+            sys.exit(1)
```
Comment (nv-hwoo) [inline model_analyzer/analyzer.py:109]: Is it okay to terminate without any message?

File: qa/L0_config/test.sh [hunk 7]
```diff
@@ -73,32 +72,24 @@ for config in ${LIST_OF_CONFIG_FILES[@]}; do
         cat $ANALYZER_LOG
         RET=1
     fi
-    MODEL_ANALYZER_ARGS="$MODEL_ANALYZER_ANALYZE_BASE_ARGS"
-    MODEL_ANALYZER_SUBCOMMAND="analyze"
-    run_analyzer
```
Comment (nv-hwoo) [inline qa/L0_config/test.sh:78]: Should we still check if the analyze subcommand works in the L0 tests since we haven't entirely deprecated it yet?

File: model_analyzer/analyzer.py (hunk 4 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py:117]: No, there is no ordering which is why I grouped them this way.

File: model_analyzer/result/result_table_manager.py (hunk 5 above)
Comment (nv-braf) [reply model_analyzer/result/result_table_manager.py:79]: Yes. Because the code below pops all the entries off the heap. Otherwise, you have to be sure that this method is the last thing you call, which I don't think is a very "clean" way of coding things.

File: qa/L0_config/test.sh (hunk 7 above)
Comment (nv-braf) [reply qa/L0_config/test.sh:78]: There's separate tests (see L0_analyze) that still use the analyze subcommand.

File: model_analyzer/analyzer.py [hunk 8]
```diff
@@ -64,7 +70,8 @@ def __init__(self, config, server, state_manager, checkpoint_required):
         self._result_manager = ResultManager(config=config,
                                              state_manager=self._state_manager)
 
-    def profile(self, client, gpus):
+    def profile(self, client: Union[TritonGRPCClient, TritonHTTPClient],
```
Comment (tgerdesnv) [inline model_analyzer/analyzer.py:73]: For the case of multiple implementations of a class, we should just type-hint the parent/interface. In this case, that is just TritonClient.

File: model_analyzer/analyzer.py [hunk 9]
```diff
@@ -98,11 +105,22 @@ def profile(self, client, gpus):
...
 
+        if (self._state_manager.exiting()):
+            sys.exit(1)
+
+        # TODO: TMA-792: This won't be needed once the Results class is used in profile
+        self._analyze_models()
+
+        self._create_summary_tables(verbose)
+
+        # TODO: TMA-790: Skip this step if the CLI option indicates the user doesn't want summary reports
```
Comment (tgerdesnv) [inline model_analyzer/analyzer.py:116]: I still think profile-only should also skip the csv output as well (which I think is 2 lines above this?)

File: model_analyzer/analyzer.py [hunk 10]
```diff
@@ -223,6 +241,42 @@ def _profile_models(self):
                 finally:
                     self._state_manager.save_checkpoint()
 
+    def _analyze_models(self):
+        # TODO: Until we get rid of analysis we need to copy some values from profile
```
Comment (tgerdesnv) [inline model_analyzer/analyzer.py:245]: Tag every TODO with a story.

File: model_analyzer/analyzer.py [hunk 11]
```diff
@@ -223,6 +241,42 @@ def _profile_models(self):
...
+
+        self._result_table_manager.create_tables()
+        self._result_table_manager.tabulate_results()
+        self._result_table_manager.export_results()
+
+        if verbose:
+            self._result_table_manager.write_results()
+
+    def _create_summary_reports(self, mode: str):
+        gpu_info = self._state_manager.get_state_variable('MetricsManager.gpus')
```
Comment (tgerdesnv) [inline model_analyzer/analyzer.py:267]: Let's avoid reaching into state manager though a hardcoded string. Ideally this goes through a public function in ResultManager, although this path seems to be MetricsManager?

File: model_analyzer/result/result_table_manager.py (hunk 5 above)
Comment (tgerdesnv) [reply model_analyzer/result/result_table_manager.py:79]: Can we not come up with a better way to iterate through sorted results without popping?

**Review by tgerdesnv — CHANGES_REQUESTED** [review]

File: qa/L0_config/test.sh (hunk 7 above)
Comment (nv-hwoo) [reply qa/L0_config/test.sh:78]: Gotcha. Thanks!

File: model_analyzer/analyzer.py (hunk 1 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py:74]: Updated to use List

File: model_analyzer/analyzer.py (hunk 6 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py:109]: Yes, the message is in the interrupt handler(). I've added a comment indicating this.

File: model_analyzer/analyzer.py (hunk 8 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py:73]: Done

File: model_analyzer/analyzer.py (hunk 10 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py:245]: Added TMA-792 tag

File: model_analyzer/analyzer.py (hunk 11 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py:267]: I've filed TMA-809 to track this work

File: model_analyzer/analyzer.py (hunk 9 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py:116]: Done

File: model_analyzer/entrypoint.py (hunk 2 above)
Comment (nv-braf) [reply model_analyzer/entrypoint.py:266]: Removed the bool wrapper

File: model_analyzer/analyzer.py (hunk 3 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py:252]: No, I've removed it. Good catch.

File: model_analyzer/analyzer.py [hunk 12]
```diff
@@ -30,6 +32,11 @@
 from .model_analyzer_exceptions \
     import TritonModelAnalyzerException
 
+from .triton.client.client import TritonClient
+from .device.gpu_device import GPUDevice
+
+from typing import Union
```
Comment (nv-hwoo) [inline model_analyzer/analyzer.py:38]: We can merge the two imports

**Review by tgerdesnv — DISMISSED** [review]

**Comment by tgerdesnv** [comment]
approved to remove my request for changes, but please don't merge until it is actually ready and fully reviewed

**Review by nv-hwoo — DISMISSED** [review]

File: model_analyzer/analyzer.py [hunk 13]
```diff
@@ -98,11 +104,22 @@ def profile(self, client, gpus):
...
+        self._analyze_models()
+
+        # TODO: TMA-790: Skip these steps if the CLI option indicates the user doesn't want summary reports
+        self._create_summary_tables(verbose)
+        self._create_summary_reports(mode)
+
         logger.info(self._get_profile_complete_string())
         logger.info("")
-        logger.info(self._get_analyze_command_help_string())
+        logger.info(self._get_report_command_help_string())
```
Comment (tgerdesnv) [inline model_analyzer/analyzer.py:120]: have we verified this still works correctly?

File: model_analyzer/record/metrics_manager.py [hunk 14]
```diff
@@ -117,7 +117,7 @@ def _init_state(self):
             if self._gpus[i].device_uuid() not in gpu_info:
                 device_info = {}
                 device = numba.cuda.list_devices()[i]
-                device_info['name'] = device.name
+                device_info['name'] = str(device.name, encoding='utf-8')
```
Comment (tgerdesnv) [inline model_analyzer/record/metrics_manager.py:120]: why was this needed?

File: model_analyzer/analyzer.py [hunk 15]
```diff
@@ -98,11 +104,22 @@ def profile(self, client, gpus):
 
         self._profile_models()
 
+        # The message is in interrupt_hanlder(), so we can just exit
```
Comment (tgerdesnv) [inline model_analyzer/analyzer.py:107]: typo

File: model_analyzer/analyzer.py (hunk 13 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py:120]: Yes, I've run profile and verified the output is correct

File: model_analyzer/record/metrics_manager.py (hunk 14 above)
Comment (nv-braf) [reply model_analyzer/record/metrics_manager.py:120]: The device name is returned as a byte-stream - this wasn't an issue before because it was stored in the checkpoint and then read back (which automatically converted it to a string when it went into json format).

File: model_analyzer/analyzer.py (hunk 15 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py]: Fixed

**Review by nv-hwoo — APPROVED** [review]
